aten/src/ATen/native/native_functions.yaml

```diff
@@ -6286,11 +6286,12 @@
   dispatch:
     CPU: legacy::cpu::_th_std
 
-- func: _amp_non_finite_check_and_unscale_(Tensor(a!) self, Tensor(b!) found_inf, Tensor inv_scale) -> ()
+- func: _amp_foreach_non_finite_check_and_unscale_(Tensor(a!)[] self, Tensor(b!) found_inf, Tensor inv_scale) -> ()
   use_c10_dispatcher: full
+  device_guard: False
```

Comment:
why does this have `device_guard: False`? (and like, _amp_update_scale doesn't).

Reply:
I'm imitating other foreach functions, all of which have `device_guard: False`.  That's a good point though, I should be more explicit about device guarding in the functions.

Reply:
Now that you mention it, existing foreach functors might be dropping the ball on device guarding as well (they should do so manually, since they're all codegenned with `device_guard: False`).  will double check those as well.

Reply:
multi_tensor_apply [guards onto the first tensor in its lists](https://github.com/pytorch/pytorch/blob/ab5edf21b0f21ab28684211390a35e2b218cf5d3/aten/src/ATen/native/cuda/MultiTensorApply.cuh#L126).
My fallback single-tensor path, for tensors MTA can't handle, uses gpu_kernel.  It's not obvious to me that gpu_kernel always guards onto its argument.  Adding explicit guard.

Also, the [tests cover](https://github.com/pytorch/pytorch/pull/44778/files#diff-e68a7bc6ba14f212e5e7eb3727394b40R1809-R1812) incoming scaled_grads on different GPUs, forcing both the MTA path and the fallback path to execute on two devices  (fallback path is forced by creating some inputs as slices so theyre not non overlapping and dense and therefore can't be handled by MTA).